pcs/git_manager.py

```diff
@@ -0,0 +1,488 @@
+            print(
+                f"\nUncommitted changes: {tracked_changes} or "
+                f"{unstaged_changes}\n"
+            )
+            error_msg = "Commit all changes before freezing"
+            if force:
+                print(f"Warning: {error_msg}")
+            else:
+                raise BadGitStateException(error_msg)
+
+    def _get_default_repo_path(self, org_name: str, project_name: str) -> Path:
+        """
+        We checkout a "default repo" to map ref names to sha1 hashes.
+        """
+        global_default_path = AOS_GLOBAL_REPOS_DIR / "_default"
+        default_repo_path = global_default_path / org_name / project_name
+        url = f"https://github.com/{org_name}/{project_name}.git"
+        cloned = self._clone_repo(url=url, clone_destination=default_repo_path)
+        if not cloned:
+            self._maybe_fetch(default_repo_path)
+        return default_repo_path
+
+    def _maybe_fetch(self, default_repo_path: Path) -> None:
+        """
+        This function periodically runs `git fetch` in the "default" repo used
+        to map ref names to sha1 hashes.  We only update if we're a new
+        process or if we haven't updated in the last half hour.
+        """
+        PROC_KEY = "process_id"
+        TIME_KEY = "last_fetch_time"
+        default_info_path = Path(str(default_repo_path) + "-info.json")
+
+        def _do_fetch():
+            with open(default_info_path, "w") as fout:
+                json.dump({PROC_KEY: os.getpid(), TIME_KEY: time.time()}, fout)
+            print(f"GitManager: fetching {default_repo_path}...")
+            porcelain.fetch(porcelain.open_repo(default_repo_path))
+
+        if not default_info_path.exists():
+            _do_fetch()
+            return
+        with open(default_info_path) as fin:
+            try:
+                info = json.load(fin)
+            except JSONDecodeError:
+                _do_fetch()
+                return
+        sec_since_last_fetch = time.time() - info[TIME_KEY]
+        if info[PROC_KEY] != os.getpid() or sec_since_last_fetch > 1800:
+            _do_fetch()
+            return
+        print(f"GitManager: NOT fetching {default_repo_path}...")
+
+    def clear_repo_cache(
+        self, repo_cache_path: Path = None, assume_yes: bool = False
+    ) -> None:
+        """
+        Completely removes all the repos that have been created or checked
+        out in the ``repo_cache_path``. Pass True to ``assume_yes`` to run
+        non-interactively.
+        """
+        repo_cache_path = repo_cache_path or AOS_GLOBAL_REPOS_DIR
+        clear_cache_path(repo_cache_path, assume_yes)
+
+    def clone_repo(
+        self, org_name: str, project_name: str, version: str
+    ) -> Path:
+        """
+        Given a GitHub org, project name, and version (a sha1 commit hash,
+        branch name, or tag name), this will clone the repo at the version
+        into the PCS/AOS cache.
+        """
+        clone_destination = (
```

Comment:
Do we want to resolve any branch or tag names to the githash (which it looks you're doing in `_checkout_version()` currently) and create a directory with that name in our local cache? Or is there a chicken and egg thing happening here where we need to checkout some version before we can use the repo to resolve the branch to a specific hash?

I guess this would be more of an optimization anyway, in the rare case where two different references point to the same underlying hash I think the current approach would just have two different directories that have the same contents.

Reply:
This is a good point, right now we're relying on the callers to resolve refs to hashes, but there's not reason not to do it right here.  Will update to not rely on the benevolence of the callers.